### PR TITLE
[JENKINS-27035] Gathering command read/write events

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-27035](https://issues.jenkins-ci.org/browse/JENKINS-27035)

Downstream of https://github.com/jenkinsci/remoting/pull/204.

* JENKINS-27035 - new API in Remoting to listen to important `Command` events
* JENKINS-49027 - Remoting 3.17: Improve reporting of JEP-200 violations in Remoting serialization
* Full changelog: https://github.com/jenkinsci/remoting/blob/master/CHANGELOG.md#317

@reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -170,7 +170,7 @@ THE SOFTWARE.
       <dependency>
         <groupId>org.jenkins-ci.main</groupId>
         <artifactId>remoting</artifactId>
-        <version>3.15-20171128.150341-4</version> <!-- TODO https://github.com/jenkinsci/remoting/pull/204 -->
+        <version>3.15-SNAPSHOT</version> <!-- TODO https://github.com/jenkinsci/remoting/pull/204 -->
```
Comment (daniel-beck) [inline pom.xml]: Huh?

Comment (jglick) [reply pom.xml]: See cab2663 commit message.

**Comment by oleg-nenashev** [comment]
on hold till the new Remoting version gets released

File: core/src/main/java/hudson/FilePath.java [hunk 2]
```diff
@@ -1159,19 +1159,22 @@ public VirtualFile toVirtualFile() {
...
+        private static final long serialVersionUID = 1L;
+        @Override
+        public Boolean invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
+            if(mkdirs(f) || f.exists())
+                return true;    // OK
 
-                // following Ant <mkdir> task to avoid possible race condition.
-                Thread.sleep(10);
+            // following Ant <mkdir> task to avoid possible race condition.
+            Thread.sleep(10);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/FilePath.java]: Uh oh, do we really have THAT in FilePath?

**Comment by oleg-nenashev** [comment]
I feel to could be moved to an API class in Remoting. E.g.

```java

class LoggerChannelListener extends ChannelListener {
    private final Logger LOGGER;

    public LoggingChannelListener() {
       ...
    }    
 
    ... // all methods
}
```

Why? It may be useful to have the same logger for Remoting CLI, Maven Plugin and other custom implementations with Channel.

WDYT @jglick ?

**Review by oleg-nenashev — APPROVED** [review]
🐝 in general, but I feel the Logging listener should be an API class in Remoting itself

**Comment by jglick** [comment]
Pushing down the logging listener is a reasonable idea, though it is a little tricky since then core would either need to attach two listeners (slightly more runtime overhead) or subclass it (somewhat poor style).

**Comment by oleg-nenashev** [comment]
I hope to release the Remoting side next week

**Comment by jglick** [comment]
Needs a few test fixes to keep up with https://github.com/jenkinsci/remoting/pull/247.

**Comment by oleg-nenashev** [comment]
It will need an integration with the master branchonce https://github.com/jenkinsci/jenkins/pull/3276 is released

**Comment by oleg-nenashev** [comment]
Crap, I forgot to merge it. #FOSDEM

**Pull request merged**